### PR TITLE
[Refactored] support undo/redo action in query builder

### PR DESCRIPTION
**Pull request by YannanGao-gs**
## Summary

support undo/redo action in query builder

improve https://github.com/finos/legend-studio/pull/2958 by encapsulating undo/redo button 

- [x] support undo/redo action in query builder
- [x] bind ctrl-z and ctrl-y
- [x] make undo/redo contextual

## How did you test this change?

- [ ] Test(s) added
- [x] Manual testing (please provide screenshots/recordings)
- [ ] No testing (please provide an explanation)

legend query:

https://github.com/finos/legend-studio/assets/73408381/b11e1d60-a673-4535-9310-61a3cd0240ec

legend studio: 

https://github.com/finos/legend-studio/assets/73408381/996829cb-3834-4bab-a91a-f7de3af31125

bind ctrl + z to undo , ctrl + y to redo

https://github.com/finos/legend-studio/assets/73408381/c2330929-54e9-46e1-a94c-74682bb9269a

make undo/redo contextual  -> ctrl + z won't close a new open modal

https://github.com/finos/legend-studio/assets/73408381/f3525a56-65e4-41a5-bdb6-80830e4db3c1

### CODE REVIEW

**Comment by YannanGao-gs** [comment]
noted: I just found out that Legend Query is not behaving consistently with Legend Studio as Legend Query's document.focus is not reset to be query builder after you close one modal

File: packages/legend-lego/src/application/RedoButton.tsx
Comment (gayathrir11) [inline packages/legend-lego/src/application/RedoButton.tsx:32]: Have you tried registering the commands in queryBuilderState instead of binding them like this?

File: packages/legend-query-builder/src/components/QueryBuilder.tsx
Comment (gayathrir11) [inline packages/legend-query-builder/src/components/QueryBuilder.tsx:389]: Is it this condition you are adding to prevent test failures `!queryBuilderState.isQuerySupported` ? What error do the tests throw?

File: packages/legend-query-builder/src/stores/QueryBuilderChangeHistoryState.ts
Comment (gayathrir11) [inline packages/legend-query-builder/src/stores/QueryBuilderChangeHistoryState.ts:94]: Do we want to handle the case where query could be unsupported? I think we shouldn't allow these undo and redo operations if query is unsupported. It's just the lambda we would have not any UI state so might just make sense to disable these operations

File: packages/legend-query-builder/src/stores/__tests__/QueryBuilderChangeDetection.test.ts
Comment (gayathrir11) [inline packages/legend-query-builder/src/stores/__tests__/QueryBuilderChangeDetection.test.ts:94]: can we please add a new test in a new file instead of modifying existing one?

**Review by gayathrir11 — DISMISSED** [review]

File: packages/legend-lego/src/application/RedoButton.tsx
Comment (YannanGao-gs) [reply packages/legend-lego/src/application/RedoButton.tsx:54]: I think maybe we could keep it for unsupported queries since we keep query snapshots no matter it's unsupported or supported in form mode

It 's currently working upon unsupported queries.

File: packages/legend-query-builder/src/stores/__tests__/QueryBuilderChangeDetection.test.ts
Comment (YannanGao-gs) [reply packages/legend-query-builder/src/stores/__tests__/QueryBuilderChangeDetection.test.ts:94]: sure! I will create a new one. I thought we could  avoid some duplicated code If we could use the existing one for query setup

File: packages/legend-query-builder/src/stores/QueryBuilderChangeHistoryState.ts
Comment (YannanGao-gs) [reply packages/legend-query-builder/src/stores/QueryBuilderChangeHistoryState.ts:94]: I think it's better to support undo/redo on unsupported queries as it might take a while to open it in the lambda editor to modify the query back to what it was before. With undo/redo, you could just hit keyboard shortcuts or click the undo button. 👀👀
However, I'm okay with making it not work with unsupported queries if others also agree on it

File: packages/legend-lego/src/application/RedoButton.tsx
Comment (YannanGao-gs) [reply packages/legend-lego/src/application/RedoButton.tsx:32]: I did! However, I found out that registerCommand() will only be initially called once. But If we want to make undo/redo contextual, we should bind undo/redo correspondingly when the condition is true or false. 
😅😅👀👀👀

File: packages/legend-lego/src/application/RedoButton.tsx
Comment (YannanGao-gs) [reply packages/legend-lego/src/application/RedoButton.tsx:35]: I did think about checking/tracking if parameters/constants/watermark/other modals are not opened, but eventually i don't think it's a good idea as I really don't want to make binding keyboard shortcuts explicitly depend on **parameters/constants/watermark/other modals** if we could just check if event.target is the root query builder.

If we are going to introduce more modals, we have to update the condition correspondingly but developers might forget to update the condition.

File: packages/legend-query-builder/src/components/QueryBuilder.tsx
Comment (YannanGao-gs) [reply packages/legend-query-builder/src/components/QueryBuilder.tsx:389]: all existing tests will fail to render result for query builder

```
  const renderResult = render(
    <ApplicationStoreProvider store={MOCK__applicationStore}>
      <TEST__BrowserEnvironmentProvider>
        <ApplicationFrameworkProvider>
          <QueryBuilder queryBuilderState={queryBuilderState} />
        </ApplicationFrameworkProvider>
      </TEST__BrowserEnvironmentProvider>
    </ApplicationStoreProvider>,
  );
```

 as this useEffect is called earlier and it will trigger queryBuilderState.buildQuery()
and will eventually fail at ..../src/stores/QueryBuilderValueSpecificationBuilder.ts line 158
```
  const _class = guaranteeNonNullable(
    queryBuilderState.class,
    'Class is required to build query',
  );
```

🫣🫣🫣🫣

File: packages/legend-lego/src/application/RedoButton.tsx
Comment (YannanGao-gs) [reply packages/legend-lego/src/application/RedoButton.tsx:32]: Yes, I did! But last time, when I tried to test this function to see if the key could be registered conditionally by turning a dummy boolean to true and false, the undo/redo behaved very weirdly. 😅 Based on John's feedback on encapsulating the undo/redo (make binding part of it), I didn't try more.
I'm open to giving it another try. Waiting for the final decision

Comment (YannanGao-gs) [reply packages/legend-lego/src/application/RedoButton.tsx:32]: I did. 👀👀😅 The current issue is it's a little hard to pass the contextual condition to queryBuilderState if we don't want to check if parameters/constants/watermark/other modals are not opened. 🥹

File: packages/legend-query-builder/src/stores/QueryBuilderChangeHistoryState.ts
Comment (gayathrir11) [reply packages/legend-query-builder/src/stores/QueryBuilderChangeHistoryState.ts:94]: @MauricioUyaguari  what do you think on this?

Comment (MauricioUyaguari) [reply packages/legend-query-builder/src/stores/QueryBuilderChangeHistoryState.ts:94]: i dont think it makes sense to support this in unsupported mode. the user doesn't see the query lambda and is not editable so if they accidently hit these shortcuts, they wont know what they are switching to/from

Comment (YannanGao-gs) [reply packages/legend-query-builder/src/stores/QueryBuilderChangeHistoryState.ts:94]: I disabled undo/redo in unsupported mode

File: packages/legend-query-builder/src/stores/__tests__/QueryBuilderChangeHistory.test.ts
Comment (MauricioUyaguari) [inline packages/legend-query-builder/src/stores/__tests__/QueryBuilderChangeHistory.test.ts:90]: This test does not really test undo/redo. 
After first undo, you should ensure you no longer see Legal Name in the projection panel. 

Additionally can you add more complicated operations. 
Maybe DND in filter. 

If the expectation is that undo/redo do not work in modals, can you test this here as well ?

File: packages/legend-query-builder/src/components/QueryBuilder.tsx
Comment (MauricioUyaguari) [inline packages/legend-query-builder/src/components/QueryBuilder.tsx:386]: `cacheNewQuery` runs `hashCode` each time. 

Have you tested the performance against larger queries (lots of columns) to ensure the UI isnt affected by this computation between each action?

File: packages/legend-query-builder/src/stores/QueryBuilderChangeHistoryState.ts
Comment (MauricioUyaguari) [inline packages/legend-query-builder/src/stores/QueryBuilderChangeHistoryState.ts:95]: could we ensure the methods here are safe to call, if this is run on every action, we should probably wrap it in a try catch and log if we fail to cached the query.

**Review by MauricioUyaguari — CHANGES_REQUESTED** [review]

File: packages/legend-query-builder/src/stores/__tests__/QueryBuilderChangeHistory.test.ts
Comment (YannanGao-gs) [reply packages/legend-query-builder/src/stores/__tests__/QueryBuilderChangeHistory.test.ts:90]: `await waitFor(() => getByText(projectionCols, 'Name'));` is checking the projection panel contains 'Name' as getByText looks for an exact match of the provided string within the text content of the elements.
 However, I could add a check to expect Legal Name to be null

File: packages/legend-lego/src/application/RedoButton.tsx
Comment (YannanGao-gs) [reply packages/legend-lego/src/application/RedoButton.tsx:35]: We are still going to bind the command keyboards for undo/redo. The discussion above is about how to bind it

File: packages/legend-query-builder/src/stores/__tests__/QueryBuilderChangeHistory.test.ts
Comment (MauricioUyaguari) [reply packages/legend-query-builder/src/stores/__tests__/QueryBuilderChangeHistory.test.ts:90]: yes but the way the current test is written, undo/redo could both be broken and the test would still pass as you are loading it with both 'Name' and 'LegalName` and then doing checks on both those columns without checking if the one who shouldnt show is actually gone.

File: packages/legend-lego/src/application/RedoButton.tsx
Comment (MauricioUyaguari) [reply packages/legend-lego/src/application/RedoButton.tsx:35]: yeah exactly we should a comment that we are not binding them via the registry command because of the detection of whether of modall is open

File: packages/legend-query-builder/src/stores/QueryBuilderChangeHistoryState.ts
Comment (MauricioUyaguari) [inline packages/legend-query-builder/src/stores/QueryBuilderChangeHistoryState.ts:121]: we should log the error not notify i think. create a new legend query. CHANGE_HISTORY_ERROR or something like that

**Comment by MauricioUyaguari** [comment]
Awesome!

**Review by MauricioUyaguari — APPROVED** [review]

**Pull request merged**
